### ISSUE TITLE
[BUG] Breaking Change in v1.25.0 release from v1.24.2

### ISSUE THREAD

**Issue by audiolion** [microsoft/playwright#16458]
https://github.com/microsoft/playwright/compare/v1.24.2...v1.25.0#diff-6070243a9a8ea57709af102355d6cc733767fd59e931c354efc75f47114f2349L30

The exported function used to be `addTestCommand`, now that has been made private, and a new function `addTestCommands` is exported, so code relying on the exported API surface is breaking on the new version.

The fix is I guess to change to use `addTestCommands` but this is not documented in the release notes, not sure if there are any differences we should be aware of.

**Comment by rwoll** [comment]
Can you include your package.json and package-lock.json? Thanks! I will be looking at this issue later, but this info will possibly be helpful here.

Leaving triaging label off for now until I circle back to look into this. Thanks!

**Comment by mxschmitt** [comment]
This usually happens when playwright-core and @olaywright/test have different versions. npm ci/npm i after aligning the versions should fix it.

**Comment by rwoll** [comment]
> Maybe I am misunderstanding something but my package.json / lock are irrelevant. 

This request is to reduce the debug loop on our end if it becomes relevant when we triage this issue later today. PW has a few packages that are part of it.

Sorry you're facing issues and thanks for hunting down the commit! I'll be looking into this in detail later today when triaging.

**Comment by mxschmitt** [comment]
> The issue arises because the recommendation is to do npx playwright test which always will grab the latest playwright core

This is not the case, `@playwright/test` will expose a bin called `playwright` so if you do `npx playwright test` it will automatically go to the `@playwright/test` NPM module which will internally call `playwright-core`. `@playwright/test`
 depends strictly on `playwright-core` with a strong version match.

We consider `addTestCommand` an implementation detail, it's not an official API which is exposed somewhere. We agree that the current behaviour is not great as well, so we should most likely add version validation to throw in the beginning if the versions diverge like afaik babel or other projects are doing.

**Comment by audiolion** [comment]
> We consider addTestCommand an implementation detail, it's not an official API which is exposed somewhere. 

Yeah for sure! We don't actually call addTestCommand, the CLI was calling it, and it seems like our CI installed the latest version of playwright (v1.25.0) and when it went to call `addTestCommands()` the function did not exist because it was finding `@playwright/test@v1.24.2` in our node_modules which exported `addTestCommand`. Given what you said about npx taking the one installed by `@playwright/test` I am not sure how the versions got mismatched, maybe some oddity of our monorepo structure or the order in which we installed things in our CI environment. Maybe `yarn dlx` behaves differently from `npx` as well.

**Comment by rwoll** [comment]
> and it seems like our CI installed the latest version of playwright (v1.25.0)

when you run the install commands on CI, is the package.json and package-lock.json in the dir tree you are running the command from? (It relates to this [request](https://github.com/microsoft/playwright/issues/16458#issuecomment-1212211694).)

For npm+npx, at least, if that will ensure expected versions of things will be run (assuming they are correct in those files). We discourage running install command globally, outside of a project:

```
╔═══════════════════════════════════════════════════════════════════════════════╗
║ WARNING: It looks like you are running 'npx playwright install' without first ║
║ installing your project's dependencies.                                       ║
║                                                                               ║
║ To avoid unexpected behavior, please install your dependencies first, and     ║
║ then run Playwright's install command:                                        ║
║                                                                               ║
║     npm install                                                               ║
║     npx playwright install                                                    ║
║                                                                               ║
║ If your project does not yet depend on Playwright, first install the          ║
║ applicable npm package (most commonly @playwright/test), and                  ║
║ then run Playwright's install command to download the browsers:               ║
║                                                                               ║
║     npm install @playwright/test                                              ║
║     npx playwright install                                                    ║
║                                                                               ║
╚═══════════════════════════════════════════════════════════════════════════════╝
```

since it's never clear what versions of what you'll be getting. (I've noted you are using yarn, so it may be slightly different, but the idea is the same: make sure your CI is set up to deterministically use PW version specified in your package.json and lockfiles.)

**Comment by rwoll** [comment]
@mxschmitt is working on better error messaging in the case of version mismatch (https://github.com/microsoft/playwright/pull/16475), but I'm trying to understand how we get into the version mismatch state, so we can add docs and/or fixes as appropriate to prevent getting into it in the first place. Thanks for your patience with all the questions :) (I don't want you to hit it again next time we release!)

**Comment by audiolion** [comment]
@rwoll I checked and though I changed this order around to first install deps, previously when this was breaking for us we were calling the command to install playwright first before installing repository deps

<img width="1027" alt="image" src="https://user-images.githubusercontent.com/2430381/184424642-938097d2-6f8e-4c28-a7b1-b886b19f72ac.png">

**Comment by rwoll** [comment]
@audiolion Any chance you can send me CI logs for the most recent time it failed? I want to better understand what commands were run and which deps were installed when. r o sswollman /@/ microsoft.com

**Comment by audiolion** [comment]
sent

**Comment by rwoll** [comment]
Thanks! @mxschmitt — forward to you, too. We can review together tomorrow

**Comment by mxschmitt** [comment]
Closing since this is not a regression or bug. It is caused when multiple `playwright-core` versions exist inside node_modules.

To fix it:

- Make sure you don't have playwright-core/playwright/playwright-chromium etc. inside package.json, `@playwright/test` is enough
- do `rm -rf node_modules` and `npm install`
- finally `npx playwright install` to install browsers

For the future we added a warning if the versions are not matching.
